### ISSUE TITLE
Bug in native callbacks

### ISSUE THREAD

**Issue by Kaosumaru** [richardeakin/Cinder-Dart#14]
Following code:

    print("1");
    customCallback( "Hey Hey" );
    print("2");
    customCallback( "Hey Hey" );
    print("3");

causes print to be called with arg "1", and then customCallback is called four times with args "Hey Hey", "2", "Hey Hey", "3".

I suspect that you expect that resolveNameHandler is called before each native callback, but it seems to be called only once for each encountered name/arity pair, and it should return new instance of Dart_NativeFunction each time.

**Comment by richardeakin** [comment]
Hm, I believe you suspect correctly, that was the assumption I made and it appears to be incorrect.

The problem is that `Dart_NativeFunction`s can only be c-style function pointers, not general-purpose std::function's or lambdas. Now, I need some other form to be able to sort what key to use when looking up in `mNativeCallbackMap`, or else go back to the way it was before std::functions were used for callbacks.  I'd rather keep the use of std::function's here if possible.. we'll see.

Thanks for reporting.

**Comment by Kaosumaru** [comment]
Typically, c-style callback give you some way to forward your context (as void *userdata, or something like that), but seems like dart api doesn't follow this convention.
I looked at it, and I have no idea if you can forward your data to Dart_NativeFunction - I think you can't.

If that's the case I guess you have two options if you want to keep std::function
1. Use debbuging api to retrieve callstack, and extract function name from it - probably slow, but it's most straightforward solution by far.
2. Use templates to provide distinct Dart_NativeFunction wrapping each std::function. This solution have some drawbacks (issues with multiple Scripts registering same lambda with different captured varaibles), but I think that can possibly be fixed. I've done some POC - here https://github.com/Kaosumaru/Cinder-Dart/blob/fix_native/src/cidart/utils/WrapCallable.h my `wrapCallable` function will return distinct Dart_NativeFunction for distinct lambdas.

**Comment by richardeakin** [comment]
I don't see any way to foward userdata to the native function either, unfortunately.

Using the stack trace does seem prohibitively slow, especially since the documentation of Dart_GetStackTrace() states the VM must first be paused.

Perhaps I am wrong, but from what I can tell your WrappCallable will return a distince Dart_NativeFunction for a distinct method signature, but not for a distinct method instance, no?

I asked for suggestions on the dart-misc mailing list ([here](https://groups.google.com/a/dartlang.org/forum/#!topic/misc/ai61iFGMM7E)), there was a promising suggestion on there but I haven't yet been able to turn this into something material:

> This is one case where I would implement a dart "CPlusFunction" class with a call method in the Dart side. Then use Dart_GetNativeInstanceField and Dart_SetNativeInstanceField to retrieve a pointer to a (C++) class that contains the std::function<...>*...

I still don't see where that CPlusFunction could be stored, and it could be retrieved from within a Dart_NativeFunction. At the same time, I don't see how we can use stateful std::function's unless there is only one (stateless) Dart_NativeFunction that handles all of them.

**Comment by Kaosumaru** [comment]
> Using the stack trace does seem prohibitively slow, especially since the documentation of  Dart_GetStackTrace() states the VM must first be paused.
Yes, that api seems to be designed for debugging. 

>Perhaps I am wrong, but from what I can tell your WrappCallable will return a distince Dart_NativeFunction for a distinct method signature, but not for a distinct method instance, no?
Well, yes. WrapCallable will produce distinct Dart_NativeFunction for lambdas, since they all have distinct type, but will fail to work with normal functions with same type signature.
I provided wrapFunction, which can be specialized with pointer to function, and thus have distinct Dart_NativeFunction for functions with same type, but well, this is a bit convoluted.

>I still don't see where that CPlusFunction could be stored, and it could be retrieved from within a Dart_NativeFunction.
Hmm. That suggestion indeed is promising. I think I know how Dart_SetNativeInstanceField is supposed to be used and what Adam Stark meant. 

I don't really know dart, so this code could not compile, but simplest implementation would be something like this:

```dart
void _registerCallback() native "RegisterCallback";
void _callCallback() native "CallCallback";

class CPlusFunction {

  CPlusFunction(String  callbackName) {
	_registerCallback(this, callbackName);
  }

   call([parameters]) {
	parameters.insert(0, this);
	return Function.apply(_callCallback, parameters);
  }

}
```

RegisterCallback in C++ gets dart_this, and callbackName, so it can do something like:
```
int callbackID = callbackNameToID(callbackName);
Dart_SetNativeInstanceField( dart_this, 0, callbackID);
```
CallCallback gets dart_this and then real arguments, so you can do:
```
int callbackID;
Dart_GetNativeInstanceField(dart_this, 0, &callbackID);
callCallbackWithID(callbackID, arguments);
```
and then, you could use it in dart:
```dart
var nativeFunction = new CPlusFunction("CallbackName");
nativeFunction(1,2,3);
```

I'm not sure, if you can call `Dart_SetNativeInstanceField` on any dart handle (since documentation for this function is "TODO(turnidge): Document."), or if it must be some magic`Dart_CreateNativeWrapperClass` class, but I guess you could alternatively just work with int.

Adam Stark is additionally suggesting to use Dart_NewWeakPersistentHandle to remove callback from C++ side when CPlusFunction is garbage collected - well, perhaps it is worth considering.

**Comment by richardeakin** [comment]
I've spent a while looking into the options around this and so far, there doesn't seem to be a general solution that will be as clean as what is in there now (and has this unfortunate bug). To make the Dart_SetNativeInstanceField() approach work, you need to call some other method that looks up the so-called 'CPlusFunction' in a map, you can't just use something like the following any longer:

```dart
void myCallback( String message ) native "myCallback";
```

I think a slightly better alternative would be to add some methods to cinder.dart that just prepend the native lookup id to the arguments, which will be used from within cidart::Script to retrieve the std::function. So, the above would become:

```dart
import "cinder.dart" as ci;

void myCallback( String message ) ci.callNative1( "myCallback", message );
```
Note the 1 suffix - as far as I can see there is no way to support variadic arguments, so for either solution, we would need something like:

```dart
void callNative0( String id ) native "cidart::callNative10";  
void callNative1( String id, var arg1 ) native "cidart::callNative1";  
void callNative2( String id, var arg1, var arg2 ) native "cidart::callNative2";  
void callNative3( String id, var arg1, var arg2, var arg3 ) native "cidart::callNative3";  
...
```

One drawback here is that the Dart_NativeArguments is always going to have an extra arg, but well that was the same for the CPlusFunction approach too.

I also proposed to the VM devs to add a void *userData that is settable from the entry resolver and retrievable from within the callback, but of course that wouldn't be anything we could use soon, even if they are willing.

**Comment by richardeakin** [comment]
Fixed with #15
